### PR TITLE
APIO-17 profile

### PR DESCRIPTION
**Pull request by nhpatt**
I have some doubts but this is the general idea of the solution.

Please comment on the things you see (and the things that are missing) and I'll make the changes to learn (that I'm quite lost at the moment).

I have to fix the tests and some things but feel free to criticize, I have so much to learn in apio and hypermedia :(

I also have a draft for alps json... but how should be served?

### CODE REVIEW

File: apio-architect-message-json-ld/src/main/java/com/liferay/apio/architect/message/json/ld/internal/JSONLDDocumentationMessageMapper.java [hunk 1]
```diff
@@ -92,6 +222,117 @@ public void onStart(
...
+			"hydra:statusCodes"
+		);
+	}
+
+	private String _getReturnValue(
+		String entity, RequestFunction requestFunction) {
+
+		String value = null;
+
+		if (HttpMethod.DELETE.equals(requestFunction.getHttpMethod())) {
```
Comment (ahdezma) [inline apio-architect-message-json-ld/src/main/java/com/liferay/apio/architect/message/json/ld/internal/JSONLDDocumentationMessageMapper.java]: Mmmmm... I'm not really sure how I feel about this...

File: apio-architect-message-json-ld/src/main/java/com/liferay/apio/architect/message/json/ld/internal/JSONLDDocumentationMessageMapper.java [hunk 2]
```diff
@@ -92,6 +222,117 @@ public void onStart(
...
+		contextBuilder.field(
+			"supportedClass"
+		).stringValue(
+			"hydra:supportedClass"
+		);
+
+		contextBuilder.field(
+			"supportedProperty"
+		).stringValue(
+			"hydra:supportedProperty"
```
Comment (ahdezma) [inline apio-architect-message-json-ld/src/main/java/com/liferay/apio/architect/message/json/ld/internal/JSONLDDocumentationMessageMapper.java]: Why aren't this values added as a constant in `JSONLDConstants`?

File: apio-architect-api/src/main/java/com/liferay/apio/architect/message/json/DocumentationMessageMapper.java [hunk 3]
```diff
@@ -50,6 +54,20 @@ public default void mapDescription(
 		JSONObjectBuilder jsonObjectBuilder, String description) {
 	}
 
+	public default void mapFormField(
+		JSONObjectBuilder jsonObjectBuilder, Object name,
+		Optional<FormField> formFieldOptional) {
+	}
+
+	public default void mapOperation(
+		JSONObjectBuilder jsonObjectBuilder, String entity,
```
Comment (ahdezma) [inline apio-architect-api/src/main/java/com/liferay/apio/architect/message/json/DocumentationMessageMapper.java]: Why is it called `entity`?

File: apio-architect-message-json-ld/src/main/java/com/liferay/apio/architect/message/json/ld/internal/JSONLDDocumentationMessageMapper.java (hunk 2 above)
Comment (nhpatt) [reply apio-architect-message-json-ld/src/main/java/com/liferay/apio/architect/message/json/ld/internal/JSONLDDocumentationMessageMapper.java]: Fixed here https://github.com/liferay/com-liferay-apio-architect/pull/192/commits/8ef7dea5b184901c8e75e930be6a8a22947ae75a

File: apio-architect-wiring-osgi-api/src/main/java/com/liferay/apio/architect/wiring/osgi/manager/representable/RepresentableManager.java [hunk 4]
```diff
@@ -31,6 +32,8 @@
 @ProviderType
 public interface RepresentableManager {
 
+	public Map<String, Representor> getRepresentorMap();
```
[the remark this answers is not in the document]
Comment (nhpatt) [reply apio-architect-wiring-osgi-api/src/main/java/com/liferay/apio/architect/wiring/osgi/manager/representable/RepresentableManager.java]: Fixed here: https://github.com/liferay/com-liferay-apio-architect/pull/192/commits/4822f071bf787a522efd1bbe38e992135be41f01

File: apio-architect-api/src/main/java/com/liferay/apio/architect/message/json/DocumentationMessageMapper.java [hunk 5]
```diff
@@ -60,4 +78,34 @@ public default void mapTitle(
 		JSONObjectBuilder jsonObjectBuilder, String title) {
 	}
 
+	public default void onFinishOperation(
+		JSONObjectBuilder jsonObjectBuilder,
```
[the remark this answers is not in the document]
Comment (nhpatt) [reply apio-architect-api/src/main/java/com/liferay/apio/architect/message/json/DocumentationMessageMapper.java]: Fixed here: https://github.com/liferay/com-liferay-apio-architect/pull/192/commits/67dfc4bf06be341ad4f7dda5590477d8ae723653

File: apio-architect-api/src/main/java/com/liferay/apio/architect/message/json/DocumentationMessageMapper.java [hunk 6]
```diff
@@ -50,6 +54,20 @@ public default void mapDescription(
 		JSONObjectBuilder jsonObjectBuilder, String description) {
 	}
 
+	public default void mapFormField(
+		JSONObjectBuilder jsonObjectBuilder, Object name,
```
[the remark this answers is not in the document]
Comment (nhpatt) [reply apio-architect-api/src/main/java/com/liferay/apio/architect/message/json/DocumentationMessageMapper.java]: Fixed here: https://github.com/liferay/com-liferay-apio-architect/pull/192/commits/aa36c9a277af69680ce557c302bb02c72b2a7a51

File: apio-architect-api/src/main/java/com/liferay/apio/architect/message/json/DocumentationMessageMapper.java [hunk 7]
```diff
@@ -50,6 +54,20 @@ public default void mapDescription(
...
+		Optional<FormField> formFieldOptional) {
+	}
+
+	public default void mapOperation(
+		JSONObjectBuilder jsonObjectBuilder, String entity,
+		RequestFunction requestFunction) {
+	}
+
+	public default void mapResource(
+		JSONObjectBuilder jsonObjectBuilder, String entity) {
```
[the remark this answers is not in the document]
Comment (nhpatt) [reply apio-architect-api/src/main/java/com/liferay/apio/architect/message/json/DocumentationMessageMapper.java]: Fixed here: https://github.com/liferay/com-liferay-apio-architect/pull/192/commits/875105b58ceee74bdccb45d2a63261df0fbe9c1d

File: apio-architect-writer-api/src/main/java/com/liferay/apio/architect/writer/DocumentationWriter.java [hunk 8]
```diff
@@ -167,6 +200,183 @@ public BuildStep requestInfo(RequestInfo requestInfo) {
 
 	}
 
+	private Optional<FormField> _findFormFieldWithName(
+		Optional<Form<FormField>> form, String name) {
```
[the remark this answers is not in the document]
Comment (nhpatt) [reply apio-architect-writer-api/src/main/java/com/liferay/apio/architect/writer/DocumentationWriter.java]: Fixed here: https://github.com/liferay/com-liferay-apio-architect/pull/192/commits/58db4abcf84e33885ab24c886c027adc0de5207c

File: apio-architect-api/src/main/java/com/liferay/apio/architect/message/json/DocumentationMessageMapper.java [hunk 9]
```diff
@@ -50,6 +54,20 @@ public default void mapDescription(
...
 	}
 
+	public default void mapFormField(
+		JSONObjectBuilder jsonObjectBuilder, Object name,
+		Optional<FormField> formFieldOptional) {
+	}
+
+	public default void mapOperation(
+		JSONObjectBuilder jsonObjectBuilder, String entity,
+		RequestFunction requestFunction) {
```
[the remark this answers is not in the document]
Comment (nhpatt) [reply apio-architect-api/src/main/java/com/liferay/apio/architect/message/json/DocumentationMessageMapper.java]: Removed, using only operations now

File: apio-architect-api/src/main/java/com/liferay/apio/architect/message/json/DocumentationMessageMapper.java [hunk 10]
```diff
@@ -60,4 +78,34 @@ public default void mapTitle(
...
+		JSONObjectBuilder propertyJsonObjectBuilder) {
+	}
+
+	public default void onFinishResource(
+		JSONObjectBuilder jsonObjectBuilder,
+		JSONObjectBuilder entityJsonObjectBuilder) {
+	}
+
+	public default void onStartOperation(
+		JSONObjectBuilder jsonObjectBuilder,
```
[the remark this answers is not in the document]
Comment (nhpatt) [reply apio-architect-api/src/main/java/com/liferay/apio/architect/message/json/DocumentationMessageMapper.java]: Passing operation

**Comment by nhpatt** [comment]
Fixed the failing test and adapted to yesterday's PR (representor with 1 parameter and FieldFunctions broke some things) //cc @ahdezma @jmnavarro

**Comment by ahdezma** [comment]
ci:test

**Pull request merged**